File: Python/13_roman_to_integer.py
class Solution(object):
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """

        '''
        my_dict = { 'I': 1, 'V': 5, 'X': 10,'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        my_twins = {'IV': 4,'IX': 9, 'XL': 40, 'XC': 90, 'CD': 400, 'CM': 900}

        my_list = list(s)
        my_sum = 0

        i = 0
        while i < len(my_list):
            if i == len(my_list) - 1:
                my_sum += my_dict[my_list[i]]
                i +=1 
            elif my_list[i] + my_list[i+1] in my_twins.keys():
                my_sum += my_twins[my_list[i] + my_list[i+1]]
                i += 2
            else:
                my_sum += my_dict[my_list[i]]
                i +=1 

        return my_sum
        '''

        d = {'I':1, 'V':5, 'X':10, 'L':50, 'C':100, 'D':500, 'M':1000}
        n = len(s)
        i = 0
        result = 0
        while i < n:
            if i < n - 1 and d[s[i]] < d[s[i + 1]]:
                result += d[s[i + 1]] - d[s[i]]
                i += 2
            else:
                result += d[s[i]]
                i += 1
        return result

File: Java/13_roman_to_integer/13_roman_to_integer.java
class Solution {
    public int romanToInt(String s) {
        int result = 0;
        for (int i=0; i<s.length()-1; i++) {
            if (s.charAt(i) == 'I' && s.charAt(i+1) == 'V') {
                i++;
                result += 4;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'I' && s.charAt(i+1) == 'X') {
                i++;
                result += 9;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'X' && s.charAt(i+1) == 'L') {
                i++;
                result += 40;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'X' && s.charAt(i+1) == 'C') {
                i++;
                result += 90;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'C' && s.charAt(i+1) == 'D') {
                i++;
                result += 400;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'C' && s.charAt(i+1) == 'M') {
                i++;
                result += 900;
                if (i == s.length()-1) return result;
            }
            else if (s.charAt(i) == 'I') result += 1;
            else if (s.charAt(i) == 'V') result += 5;
            else if (s.charAt(i) == 'X') result += 10;
            else if (s.charAt(i) == 'L') result += 50;
            else if (s.charAt(i) == 'C') result += 100;
            else if (s.charAt(i) == 'D') result += 500;
            else if (s.charAt(i) == 'M') result += 1000;
        }
        if (s.charAt(s.length()-1) == 'I') result += 1;
        else if (s.charAt(s.length()-1) == 'V') result += 5;
        else if (s.charAt(s.length()-1) == 'X') result += 10;
        else if (s.charAt(s.length()-1) == 'L') result += 50;
        else if (s.charAt(s.length()-1) == 'C') result += 100;
        else if (s.charAt(s.length()-1) == 'D') result += 500;
        else if (s.charAt(s.length()-1) == 'M') result += 1000;

        return result;
    }
}